### PR TITLE
Update typing and checks with mypy for settings module

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

- Start add typing checks to codebase. For now `napari.utils.settings`
- Also add a scheduled workflow to check this specific module with mypy.

We can start adding modules and submodules, little by little.

## Type of change

- [x] Task

# References

See #2751

# How has this been tested?

- [x] example: all tests pass with my change

### CODE REVIEW

File: MANIFEST.in
Comment (tlambert03) [inline MANIFEST.in]: would probably add this to `[tool.check-manifest][ignore]` in pyproject.toml instead of shipping it with the dist

File: .github/workflows/test_typing.yml
Comment (tlambert03) [inline .github/workflows/test_typing.yml:6]: cron feels a bit odd here (in terms of preventing new issues from creeping in)... is this just to make it less annoying for contributors?

File: .github/workflows/test_typing.yml
Comment (tlambert03) [inline .github/workflows/test_typing.yml]:
```suggestion
  typing:
```
?

File: .github/workflows/test_typing.yml
Comment (tlambert03) [inline .github/workflows/test_typing.yml]:
```suggestion
          pip install mypy
```

File: napari/utils/events/evented_model.py
Comment (tlambert03) [inline napari/utils/events/evented_model.py]: ?

**Review by tlambert03 — APPROVED** [review]
This is a fine start! :)

left a few comments.  Mostly, I'm wondering how/when to trigger this.  Cron job feels a bit odd... (I'm guessing it's to make it easier on contributors?)

Another thing that we could do is add it to pre-commit with _just_ the modules that pass checks, and gradually add to them:

```
-   repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.812
    hooks:
    - id: mypy
      additional_dependencies: [pydantic]
      files: napari/utils/settings
      # or files: napari/utils/settings|napari/other/|napari/another
```

File: .github/workflows/test_typing.yml
Comment (goanpeca) [reply .github/workflows/test_typing.yml:6]: > is this just to make it less annoying for contributors?

Correct! But the pre-commit ideas works, but mypy is kinda slow :-\

File: .github/workflows/test_typing.yml
Comment (goanpeca) [reply .github/workflows/test_typing.yml]: Whoops

File: .github/workflows/test_typing.yml
Comment (goanpeca) [reply .github/workflows/test_typing.yml]: 🤦🏼

File: napari/utils/events/evented_model.py
Comment (goanpeca) [reply napari/utils/events/evented_model.py]: Mypy was complaining about this :-\

File: .github/workflows/test_typing.yml
Comment (tlambert03) [reply .github/workflows/test_typing.yml:6]: yeah, but I don't think it will check the whole repo right?  I'm actually not sure here... it's possible that by using the `files` argument (to restrict the scope of the checks) we might actually be forcing it to run them on all of those files every time?  (even if the specific commit didn't touch anything in there)

File: .github/workflows/test_typing.yml
Comment (tlambert03) [reply .github/workflows/test_typing.yml]: it's funny cause mypi is a thing too :)

File: MANIFEST.in
Comment (goanpeca) [reply MANIFEST.in]: Fixed

File: .github/workflows/test_typing.yml
Comment (goanpeca) [reply .github/workflows/test_typing.yml:6]: >actually be forcing it to run them on all of those files every time? (even if the specific commit didn't touch anything in there)

I can check, but yeah in principle I think it would check them all the time.

File: napari/utils/events/evented_model.py
Comment (goanpeca) [reply napari/utils/events/evented_model.py]: Fixed!

File: .github/workflows/test_typing.yml
Comment (sofroniewn) [inline .github/workflows/test_typing.yml:25]: I guess with time this will gradually grow!

**Review by sofroniewn — APPROVED** [review]
LGTM! I like the changes. We've got a couple merge conflicts, and i'll let others weigh in again before merge, but in general I like the step in this direction. Thanks all!!

File: .github/workflows/test_typing.yml
Comment (goanpeca) [reply .github/workflows/test_typing.yml:25]: We could try to keep adding things slowly, one module per week or so, so we can slowly but steadily move in this direction :)

File: .github/workflows/test_typing.yml
Comment (sofroniewn) [reply .github/workflows/test_typing.yml:6]: Is this discussion resolved? Can I merge now, or do we want to change before merge?

Comment (tlambert03) [reply .github/workflows/test_typing.yml:6]: I don't mind moving forward with this for now.  So I'll resolve this.  I'm not sure who is going to be monitoring these cron jobs, so there's a vague concern that we'll get regressions if we don't put this into PR tests eventually.  and I'm actually ok with the contributor burden of enforcing typing on PRs for sections of the code that we have already carefully fixed (as long as it's only those files)

Comment (sofroniewn) [reply .github/workflows/test_typing.yml:6]: Yeah, agreed - i think putting into tests eventually for things we have fixed will be right move, will help prevent regressions. can do that in followup PRs

**Comment by sofroniewn** [comment]
Ok, I think this will be good to go then once tests pass. Thanks all!!

**Review by sofroniewn — APPROVED** [review]

**Comment by sofroniewn** [comment]
This now has quite a few conflicts, sorry! If you can update this one more time @goanpeca I'll then merge

**Comment by goanpeca** [comment]
>This now has quite a few conflicts, sorry! If you can update this one more time @goanpeca I'll then merge

Fixed!

**Pull request merged**
